build/scripts/build.ps1

```diff
@@ -1,84 +1,401 @@
+    # that we do not reuse MSBuild nodes from other jobs/builds on the machine. Otherwise,
+    # we'll run into issues such as https://github.com/dotnet/roslyn/issues/6211.
+    # MSBuildAdditionalCommandLineArgs=
+    $args = "/p:TreatWarningsAsErrors=true /warnaserror /nologo /m /nodeReuse:false /consoleloggerparameters:Verbosity=minimal";
+    
+    if ($logFile -ne "") {
+        $args += " /filelogger /fileloggerparameters:Verbosity=normal;logFile=$logFile";
     }
-    Exec-Command $msbuild $args | Out-Host
+
+    if ($cibuild) { 
+        $args += " /p:PathMap=`"$($repoDir)=q:\roslyn`" /p:Feature=pdb-path-determinism" 
+    }
+
+    $args += " $buildArgs"
+    Exec-Command $msbuild $args
 }
 
-try {
-    if ($badArgs -ne $null) {
-        Write-Host "Bad arguments: $badArgs"
-        Print-Usage
-        exit 1
+# Create a bootstrap build of the compiler.  Returns the directory where the bootstrap buil 
+# is located. 
+#
+# Important to not set $script:bootstrapDir here yet as we're actually in the process of 
+# building the bootstrap.
+function Make-BootstrapBuild() {
+
+    $bootstrapLog = Join-Path $binariesDir "Bootstrap.log"
+    Write-Host "Building Bootstrap compiler"
+    Run-MSBuild "/p:UseShippingAssemblyVersion=true /p:InitialDefineConstants=BOOTSTRAP build\Toolset\Toolset.csproj /p:Configuration=$buildConfiguration" -logFile $bootstrapLog | Out-Host
+    $dir = Join-Path $binariesDir "Bootstrap"
+    Remove-Item -re $dir -ErrorAction SilentlyContinue
+    Create-Directory $dir
+    Move-Item "$configDir\Exes\Toolset\*" $dir
+
+    Write-Host "Cleaning Bootstrap compiler artifacts"
+    Run-MSBuild "/t:Clean build\Toolset\Toolset.csproj /p:Configuration=$buildConfiguration" | Out-Host 
+    Stop-BuildProcesses
+    return $dir
+}
+
+function Test-PerfCorrectness() {
+    Run-MSBuild "Roslyn.sln /p:Configuration=$buildConfiguration /p:DeployExtension=false"
+    Exec-Block { & ".\Binaries\$buildConfiguration\Exes\Perf.Runner\Roslyn.Test.Performance.Runner.exe" --ci-test } | Out-Host
+}
+
+function Test-PerfRun() { 
+    Run-MSBuild "Roslyn.sln /p:Configuration=$buildConfiguration /p:DeployExtension=false"
+
+    # Check if we have credentials to upload to benchview
+    $extraArgs = @()
+    if ((Test-Path env:\GIT_BRANCH) -and (Test-Path env:\BV_UPLOAD_SAS_TOKEN)) {
+        $extraArgs += "--report-benchview"
+        $extraArgs += "--branch=$env:GIT_BRANCH"
+
+        # Check if we are in a PR or this is a rolling submission
+        if (Test-Path env:\ghprbPullTitle) {
+            $submissionName = $env:ghprbPullTitle.Replace(" ", "_")
+            $extraArgs += "--benchview-submission-name=""$submissionName"""
+            $extraArgs += "--benchview-submission-type=private"
+        } 
+        else {
+            $extraArgs += "--benchview-submission-type=rolling"
+        }
+
+        Create-Directory ".\Binaries\$buildConfiguration\tools\"
```

Comment:
Working directory is always the repo root in this script. Look at the push-location calls.